pkg/autoupdate/tuf/library_manager.go

```diff
@@ -0,0 +1,349 @@
+	}
+
+	// Ensure our staged updates and updates directories exist
+	for _, binary := range binaries {
+		// Create the directory for staging updates
+		if err := os.MkdirAll(ulm.stagedUpdatesDirectory(binary), 0755); err != nil {
+			return nil, fmt.Errorf("could not make staged updates directory for %s: %w", binary, err)
+		}
+
+		// Create the update library
+		if err := os.MkdirAll(ulm.updatesDirectory(binary), 0755); err != nil {
+			return nil, fmt.Errorf("could not make updates directory for %s: %w", binary, err)
+		}
+	}
+
+	return &ulm, nil
+}
+
+// updatesDirectory returns the update library location for the given binary.
+func (ulm *updateLibraryManager) updatesDirectory(binary autoupdatableBinary) string {
+	return filepath.Join(ulm.baseDir, string(binary))
+}
+
+// stagedUpdatesDirectory returns the location for staged updates -- i.e. updates
+// that have been downloaded, but not yet verified.
+func (ulm *updateLibraryManager) stagedUpdatesDirectory(binary autoupdatableBinary) string {
+	return filepath.Join(ulm.baseDir, fmt.Sprintf("%s-staged", binary))
+}
+
+// addToLibrary adds the given target file to the library for the given binary,
+// downloading and verifying it if it's not already there.
+func (ulm *updateLibraryManager) AddToLibrary(binary autoupdatableBinary, targetFilename string, targetMetadata data.TargetFileMeta) error {
+	// Check to see if the current running version is the version we were requested to add;
+	// return early if it is, but don't error out if we can't determine the current version.
+	currentVersion, err := ulm.currentRunningVersion(binary)
+	if err != nil {
+		level.Debug(ulm.logger).Log("msg", "could not get current running version", "binary", binary, "err", err)
+	} else if currentVersion == ulm.versionFromTarget(binary, targetFilename) {
+		// We don't need to download the current running version because it already exists,
+		// either in this updates library or in the original install location.
+		return nil
+	}
+
+	if ulm.alreadyAdded(binary, targetFilename) {
+		return nil
+	}
+
+	// Remove downloaded archives after update, regardless of success
+	defer ulm.tidyStagedUpdates(binary)
+
+	stagedUpdatePath, err := ulm.stageAndVerifyUpdate(binary, targetFilename, targetMetadata)
+	if err != nil {
+		return fmt.Errorf("could not stage update: %w", err)
+	}
+
+	if err := ulm.moveVerifiedUpdate(binary, targetFilename, stagedUpdatePath); err != nil {
+		return fmt.Errorf("could not move verified update: %w", err)
+	}
+
+	return nil
+}
+
+// alreadyAdded checks if the given target already exists in the update library.
+func (ulm *updateLibraryManager) alreadyAdded(binary autoupdatableBinary, targetFilename string) bool {
+	updateDirectory := filepath.Join(ulm.updatesDirectory(binary), ulm.versionFromTarget(binary, targetFilename))
+
+	return autoupdate.CheckExecutable(context.TODO(), executableLocation(updateDirectory, binary), "--version") == nil
```

Comment:
Does this actually work on windows given the weird GUI output issues we had?

Reply:
It's hard to tell because when testing, the running version will always be a local dirty build that doesn't match what TUF has. (I do know the check worked correctly for osquery on Windows but I don't think that was affected by the same issue?)

But this is reusing the same check from the autoupdate package, so it should at least be consistent with what we have currently?